File: go/916.go
package main

func maximu(a, b int) int {
	if a > b {
		return a
	}

	return b
}

func wordSubsets(words1 []string, words2 []string) []string {
	freqA := make([][]int, len(words1))
	uniFreq := make([]int, 26)

	for i := 0; i < len(words1); i++ {
		freqA[i] = make([]int, 26)
	}

	for i := 0; i < len(words1); i++ {
		for j := 0; j < len(words1[i]); j++ {
			freqA[i][words1[i][j]-'a']++
		}
	}

	for i := 0; i < len(words2); i++ {
		f := make([]int, 26)

		for j := 0; j < len(words2[i]); j++ {
			f[words2[i][j]-'a']++
		}

		for j := 0; j < 26; j++ {
			uniFreq[j] = maximu(uniFreq[j], f[j])
		}
	}

	result := []string{}

	for i := 0; i < len(words1); i++ {
		flag := true

		for k := 0; k < 26 && flag; k++ {
			if freqA[i][k] < uniFreq[k] {
				flag = false
			}
		}

		if flag {
			result = append(result, words1[i])
		}
	}

	return result
}

File: zig/916.zig
const std = @import("std");

fn solution(allocator: std.mem.Allocator, words1: []const []const u8, words2: []const []const u8) ![][]const u8 {
    var result: std.ArrayList([]const u8) = .empty;

    var freq = try allocator.alloc([]i32, words1.len);
    for (0..words1.len) |i| {
        freq[i] = try allocator.alloc(i32, 26);
        @memset(freq[i], 0);
    }

    for (0..words1.len) |i| {
        for (words1[i]) |c| {
            freq[i][c - 'a'] += 1;
        }
    }

    var uniFreq = try allocator.alloc(i32, 26);
    @memset(uniFreq, 0);
    defer {
        result.deinit(allocator);
        allocator.free(uniFreq);

        for (0..words1.len) |i| {
            allocator.free(freq[i]);
        }

        allocator.free(freq);
    }

    for (words2) |word| {
        var f = try allocator.alloc(i32, 26);
        defer allocator.free(f);
        @memset(f, 0);

        for (word) |c| {
            f[c - 'a'] += 1;
        }

        for (0..26) |i| {
            uniFreq[i] = @max(uniFreq[i], f[i]);
        }
    }

    for (0..words1.len) |i| {
        var flag: bool = true;

        for (0..26) |j| {
            if (freq[i][j] < uniFreq[j]) {
                flag = false;
                break;
            }
        }

        if (flag) {
            try result.append(allocator, words1[i]);
        }
    }

    return result.toOwnedSlice(allocator);
}

const expect = std.testing.expect;

test "test case #1" {
    const allocator = std.testing.allocator;

    const words1 = [_][]const u8{ "amazon", "apple", "facebook", "google", "leetcode" };
    const words2 = [_][]const u8{ "e", "o" };

    const expected = [_][]const u8{ "facebook", "google", "leetcode" };
    const actual = try solution(allocator, &words1, &words2);
    defer allocator.free(actual);

    try expect(expected.len == actual.len);
    for (0..expected.len) |i| {
        try expect(std.mem.eql(u8, expected[i], actual[i]));
    }
}

test "test case #2" {
    const allocator = std.testing.allocator;

    const words1 = [_][]const u8{ "amazon", "apple", "facebook", "google", "leetcode" };
    const words2 = [_][]const u8{ "lc", "eo" };

    const expected = [_][]const u8{"leetcode"};
    const actual = try solution(allocator, &words1, &words2);
    defer allocator.free(actual);

    try expect(expected.len == actual.len);
    for (0..expected.len) |i| {
        try expect(std.mem.eql(u8, expected[i], actual[i]));
    }
}

test "test case #3" {
    const allocator = std.testing.allocator;

    const words1 = [_][]const u8{ "acaac", "cccbb", "aacbb", "caacc", "bcbbb" };
    const words2 = [_][]const u8{ "c", "cc", "b" };

    const expected = [_][]const u8{"cccbb"};
    const actual = try solution(allocator, &words1, &words2);
    defer allocator.free(actual);

    try expect(expected.len == actual.len);
    for (0..expected.len) |i| {
        try expect(std.mem.eql(u8, expected[i], actual[i]));
    }
}
